modules/fvad/fvad.c

```diff
@@ -0,0 +1,365 @@
+/**
+ * @file fvad.c  Voice Activity Detection
+ *
+ * Uses libfvad from https://github.com/dpirch/libfvad
+ *
+ * Copyright (C) 2023 Lars Immisch
+ */
+#include <string.h>
+#include <stdlib.h>
+#include <re.h>
+#include <rem.h>
+#include <baresip.h>
+#include <fvad.h>
+#include "fvad.h"
+
+
+/**
+ * @defgroup fvad fvad
+ *
+ * Voice Activity Detection for the audio-signal.
+ *
+ * It is using the aufilt API to get the audio samples.
+ */
+
+
+struct vad_enc {
+	struct aufilt_enc_st af;  /* inheritance */
+	struct le le;
+	const struct audio *au;
+	bool vad_tx;
+	Fvad *fvad;
+	const struct call *call;
```

Comment:
You can set the pointer to the call object in the first `..._update()` function. If the pointer is already set, then you don't need another search for the call. Note, that you should not assume any order between the call of `encode_update()` and `decode_update()`.

Then you don't need the pointer to audio.

Reply:
I don't see how I can do that. The `..._update` functions receive an `audio` pointer, but not a `call` pointer.

So the app needs to call `uag_filter_calls` when it first receives the first `UA_EVENT_MODULE` event. The search functions will then set the `call` pointer and this will be filled in on subsequent invocations of `module_event`.